services/cloudapi/cloudapiserver/server.go

```diff
@@ -0,0 +1,211 @@
+			return fmt.Errorf("cannot shut down a cube server, set to %s instead", CUBE_VMSTATE_STOP)
+		}
+		if newVmState == "" {
+			return fmt.Errorf("cannot update a suspended Cube Server, must change the state to %s first", CUBE_VMSTATE_START)
+		}
+		if strings.EqualFold(newVmState, CUBE_VMSTATE_START) && strings.EqualFold(currentVmState, CUBE_VMSTATE_STOP) {
+			return fs.Start(ctx, datacenterID, serverID, serverType)
+		}
+		if strings.EqualFold(newVmState, CUBE_VMSTATE_STOP) && strings.EqualFold(currentVmState, CUBE_VMSTATE_START) {
+			return SuspendCubeLast
+		}
+
+	}
+
+	return nil
+}
+
+func (fs *Service) GetVmState(ctx context.Context, datacenterID, serverID string) (string, error) {
+	server, err := fs.FindById(ctx, datacenterID, serverID, 0)
+	if err != nil {
+		return "", err
+	}
+	return *server.Properties.VmState, nil
+}
+
+func (fs *Service) GetServerType(ctx context.Context, datacenterID, serverID string) (string, error) {
+	server, err := fs.FindById(ctx, datacenterID, serverID, 0)
+	if err != nil {
+		return "", err
+	}
+	return *server.Properties.Type, nil
+}
+
+func (fs *Service) Start(ctx context.Context, datacenterID, serverID, serverType string) error {
+
+	switch serverType {
+
+	case ENTERPRISE_SERVER_TYPE:
+		apiResponse, err := fs.Client.ServersApi.DatacentersServersStartPost(ctx, datacenterID, serverID).Execute()
+		apiResponse.LogInfo()
+		if err != nil {
+			return err
+		}
+		return utils.WaitForResourceToBeReady(ctx, fs.D, fs.checkExpectedVmStateFn(ctx, datacenterID, ENTERPRISE_VMSTATE_START))
+
+	case CUBE_SERVER_TYPE:
+		apiResponse, err := fs.Client.ServersApi.DatacentersServersResumePost(ctx, datacenterID, serverID).Execute()
+		apiResponse.LogInfo()
+		if err != nil {
+			return err
+		}
+		return utils.WaitForResourceToBeReady(ctx, fs.D, fs.checkExpectedVmStateFn(ctx, datacenterID, ENTERPRISE_VMSTATE_START))
+	}
+
+	return fmt.Errorf("cannot start unknown server type: %s", serverType)
+
+}
+
+func (fs *Service) Stop(ctx context.Context, datacenterID, serverID, serverType string) error {
+
+	switch serverType {
+
+	case ENTERPRISE_SERVER_TYPE:
+		apiResponse, err := fs.Client.ServersApi.DatacentersServersStopPost(ctx, datacenterID, serverID).Execute()
+		apiResponse.LogInfo()
+		if err != nil {
+			return err
+		}
+		return utils.WaitForResourceToBeReady(ctx, fs.D, fs.checkExpectedVmStateFn(ctx, datacenterID, ENTERPRISE_VMSTATE_STOP))
+
+	case CUBE_SERVER_TYPE:
+		apiResponse, err := fs.Client.ServersApi.DatacentersServersSuspendPost(ctx, datacenterID, serverID).Execute()
+		apiResponse.LogInfo()
+		if err != nil {
+			return err
+		}
+		return utils.WaitForResourceToBeReady(ctx, fs.D, fs.checkExpectedVmStateFn(ctx, datacenterID, CUBE_VMSTATE_STOP))
+	}
+
+	return fmt.Errorf("cannot stop unknown server type: %s", serverType)
+
+}
+
+func (fs *Service) checkExpectedVmStateFn(ctx context.Context, dcId, expectedState string) func(ctx context.Context, d *schema.ResourceData) (bool, error) {
```

Comment:
add a todo here that we should just change the function sig that is received by the interface.

Reply:
updated function signature to return a ResourceReadyFunc